docs/nebula-design-concepts.md

```diff
@@ -0,0 +1,216 @@
+
+- Parallel Raft
+
+      Partitions of same ID from multiple machines form a raft group. And the concurrent operations are implemented with multiple sets of Raft groups.
+
+- Write Path & batch
+
+      In Raft protocol, multi-machine synchronization depends on log ID orders, which
+  lead to low throughput. Nebula achieves high throughput by batch and out-of-order
+  submission.
+
+- Copy learner based on synchronism
+
+      When added to cluster, new machines will be tagged as learner and pull data
+  from leader/follower synchronically. The learner will be tagged as follower and
+  participate in Raft protocol when it catches up leader.
+
+- Load-balance
+
+      For machines with high stress, migrate the partition they serve to machines
+  with  low stress for better load balancing.
+
+![image][example6]
+
+### Metadata service
+
+The process corresponding to the Meta service is nebula-metad, and its main functions are:
+
+-  User management
+
+      Nebula Graph's user system includes God user, Admin, User, and Guest, each is different in operating privileges.
+
+- Cluster configuration management
+
+      Support online and offline new servers.
+
+- Graph space management
+
+      Add, drop graph space, modify graph space configuration (Raft copies)
+
+-   Schema management
+
+      Nebula Graph is a strong schema database.
+
+      - Types of Tag and Edge properties are recorded by Meta service. Supported types are: int, double, timestamp, list, etc.
+
+      - Multi-version management, supporting adding, modifying and deleting schema, and recording its version.
+
+      - TTL (time-to-live) management, supporting automatic data deletion and space reclamation.
+
+The Meta Service layer is state, and just like the Storage layer, it maintains state persistence through KV Store.
+
+![image][example7]
+
+### Query Engine & Query Language (nGQL)
+
+The process corresponding to the computation layer is nebula-graphd, which consists
+of computing vertices that are fully equal, stateless, and unrelated. There is no
+communication among the computing vertices.
+
+The main function of the Query Engine layer is to analysis the nGQL text sent by the client and generate an execution plan through lexical analysis (Lexer) and parsing (Parser). Then the execution plan will pass to the the execution engine. The execution engine obtains the schema of the graph vertices and edges through the Meta Service, then obtains the data of the vertices and edges through the storage engine layer.
+
+The main optimizations of the Query Engine layer are:
+
+- Asynchronous and concurrent execution
+
+      Since both IO and network are long-latency operations, asynchronous and concurrent operations are required. In addition, to avoid a single long query affecting subsequent queries, a separate resource pool is set for each query to guarantee quality of service (QoS).
+
```

Comment:
_concurrent execution_ or _parallel execution_ ?
I/O operations and network transmission are time-consuming. Asynchronous and Parallel operations are adopted in Nebula Query Engine to guarantee QoS. Also, a separate resource pool is set for each query to avoid the long tail effect of those time-costly queries.

Reply:
> _concurrent execution_ or _parallel execution_ ?
> I/O operations and network transmission are time-consuming. Asynchronous and Parallel operations are adopted in Nebula Query Engine to guarantee QoS. Also, a separate resource pool is set for each query to avoid the long tail effect of those time-costly queries.

Correct.  parallel